src/oidcendpoint/oidc/token_coop.py

```diff
@@ -131,82 +152,121 @@ def _access_token(self, req, **kwargs):
-            keyjar = ""
+HELPER_BY_GRANT_TYPE = {
+    "authorization_code": AccessToken,
+    "refresh_token": RefreshToken,
+}
 
-        request.verify(keyjar=keyjar, opponent_id=client_id)
 
-        if "client_id" not in request:  # Optional for refresh access token request
-            request["client_id"] = client_id
+class TokenCoop(Endpoint):
+    request_cls = oidc.Message
+    response_cls = oidc.AccessTokenResponse
+    error_cls = TokenErrorResponse
+    request_format = "json"
+    request_placement = "body"
+    response_format = "json"
+    response_placement = "body"
+    endpoint_name = "token_endpoint"
+    name = "token"
+    default_capabilities = {"token_endpoint_auth_signing_alg_values_supported": None}
 
-        logger.debug("%s: %s" % (request.__class__.__name__, sanitize(request)))
+    def __init__(self, endpoint_context, new_refresh_token=False, **kwargs):
+        Endpoint.__init__(self, endpoint_context, **kwargs)
+        self.post_parse_request.append(self._post_parse_request)
+        if "client_authn_method" in kwargs:
+            self.endpoint_info["token_endpoint_auth_methods_supported"] = kwargs[
+                "client_authn_method"
+            ]
 
-        return request
+        # self.allow_refresh = False
+        self.new_refresh_token = new_refresh_token
 
-    def _post_parse_request(self, request, client_id="", **kwargs):
-        if request["grant_type"] == "authorization_code":
-            return self._access_token_post_parse_request(request, client_id, **kwargs)
-        else:  # request["grant_type"] == "refresh_token":
-            if self.allow_refresh:
-                return self._refresh_token_post_parse_request(
-                    request, client_id, **kwargs
+        self.configure_grant_types(kwargs.get("grant_types_supported"))
+
+    def configure_grant_types(self, grant_types_supported):
+        if grant_types_supported is None:
+            self.helper = {k: v(self) for k, v in HELPER_BY_GRANT_TYPE.items()}
+            return
+
+        self.helper = {}
+        # TODO: do we want to allow any grant_type?
+        for grant_type, grant_type_options in grant_types_supported.items():
+            if (
+                    grant_type_options in ("default", None)
+                    and grant_type in HELPER_BY_GRANT_TYPE
+            ):
+                self.helper[grant_type] = HELPER_BY_GRANT_TYPE[grant_type]
+                continue
+            _conf = grant_type_options.get('kwargs', {})
+            try:
+                if isinstance(grant_type_options["class"], str):
+                    grant_class = importer(grant_type_options["class"])
+                else:
+                    grant_class = grant_type_options["class"]
+                self.helper[grant_type] = grant_class(self, _conf)
+            except KeyError:
+                raise ProcessError(
+                        "Grant type is invalid or missing a valid class to import."
                 )
```

Comment:
Improve error message?

Check if the required fields have been initialized and then initiate the class

Reply:
OK I broke this part in multiple checks and exception handers.